### PR TITLE
Implement property test for add/remove_liquidity() and exchange().

### PR DESCRIPTION
**Pull request by vivekvpandya**
Also fix minor bug found with proptesting.

### CODE REVIEW

File: frame/curve-amm/src/tests.rs [hunk 1]
```diff
@@ -3,12 +3,45 @@ use frame_support::assert_ok;
...
+use proptest::prelude::*;
 use sp_runtime::{
+	helpers_128bit::multiply_by_rational,
 	traits::{Saturating, Zero},
 	FixedPointNumber, FixedU128, Permill,
 };
 use sp_std::cmp::Ordering;
 
+/// Accepts -2, -1, 0, 1, 2
+macro_rules! prop_assert_zero_epsilon {
```
Comment (dzmitry-lahoda) [inline frame/curve-amm/src/tests.rs:15]: afaik @hussein-aitlahcen moved that to reusable codes

File: frame/curve-amm/src/tests.rs [hunk 2]
```diff
@@ -229,3 +262,126 @@ fn exchange_test() {
 		assert!(65000 - Tokens::balance(MockCurrencyId::USDC, &CHARLIE) < 10);
 	});
 }
+
+proptest! {
+	#![proptest_config(ProptestConfig::with_cases(10000))]
+
+	#[test]
+	fn proptest_exchange(alice_balance in 1..u32::MAX,
```
Comment (dzmitry-lahoda) [inline frame/curve-amm/src/tests.rs:270]: good,so test to test that curve is balanced would be awesome (kind of if make inbalance, interset grows fast or like)

**Review by dzmitry-lahoda — DISMISSED** [review]

**Review by dzmitry-lahoda — APPROVED** [review]

File: frame/curve-amm/src/tests.rs (hunk 1 above)
Comment (hussein-aitlahcen) [reply frame/curve-amm/src/tests.rs:15]: Not yet in main sadly

File: frame/curve-amm/src/tests.rs [hunk 3]
```diff
@@ -229,3 +262,126 @@ fn exchange_test() {
...
+		assert_eq!(Tokens::balance(MockCurrencyId::USDC, &ALICE), 0);
+		let pool = CurveAmm::pool(pool_id);
+		assert!(pool.is_some());
+		let bob_amounts = vec![bob_balance as u128, bob_balance as u128];
+		assert_ok!(CurveAmm::add_liquidity(&BOB, pool_id, bob_amounts.clone(), 0u128));
+		let bob_lp_balance = Tokens::balance(pool_lp_asset, &BOB);
+		assert_ne!(bob_balance, 0);
+		assert_eq!(Tokens::balance(MockCurrencyId::USDT, &BOB), 0);
+		assert_eq!(Tokens::balance(MockCurrencyId::USDC, &BOB), 0);
+		let min_amt = vec![0u128, 0u128];
```
Comment (KaiserKarel) [inline frame/curve-amm/src/tests.rs:369]: Only nitpick I have here is that this is very difficult to follow, if you could create groups like this, and explain what you are checking/what stage of the procedure we're in

``
let a = 1;

// Test that arithmetic works
{
   assert_eq!(1, 1);
}

// Test that it works for zero
{ 
   assert_neq!(a, 0);
}
```

**Review by KaiserKarel — CHANGES_REQUESTED** [review]
If you could add some comments in the very big test, perhaps even grouping stages, that'd be great

File: frame/curve-amm/src/tests.rs (hunk 2 above)
Comment (vivekvpandya) [reply frame/curve-amm/src/tests.rs:270]: sure, I have noted this. I will try it in next PR.

File: frame/curve-amm/src/tests.rs [hunk 4]
```diff
@@ -229,3 +262,149 @@ fn exchange_test() {
...
+		assert_ok!(CurveAmm::exchange(&CHARLIE, pool_id, 1, 0, alice_balance as u128, 0));
+		prop_assert_epsilon!(alice_balance as u128, Tokens::balance(MockCurrencyId::USDC, &CHARLIE));
+		Ok(())
+	}).unwrap();
+	}
+
+	#[test]
+	fn proptest_add_remove_liquidity(alice_balance in 0..u32::MAX,
+									 bob_balance in 0..u32::MAX) {
+	new_test_ext().execute_with(|| {
```
Comment (KaiserKarel) [inline frame/curve-amm/src/tests.rs:339]: weirdly formatted?

**Review by KaiserKarel — APPROVED** [review]

File: frame/curve-amm/src/tests.rs (hunk 4 above)
Comment (vivekvpandya) [reply frame/curve-amm/src/tests.rs:339]: On my system vim and vscode shows this line properly aligned.

**Review by dzmitry-lahoda — APPROVED** [review]

**Pull request merged**